net/libfab/libfab.c

```diff
@@ -940,33 +958,36 @@ static void libfab_txep_comp_read(struct fid_cq *cq, struct m0_fab__tm *tm)
  */
 static void libfab_poller(struct m0_fab__tm *tm)
 {
+	struct m0_net_end_point  *net;
 	struct m0_fab__ev_ctx    *ctx;
 	struct m0_fab__ep        *xep;
 	struct m0_fab__active_ep *aep;
 	struct fid_cq            *cq;
 	struct epoll_event        ev;
 	int                       ev_cnt;
 	int                       ret;
+	int                       err;
 
 	libfab_tm_event_post(tm, M0_NET_TM_STARTED);
 	while (tm->ftm_state != FAB_TM_SHUTDOWN) {
+		ret = fi_trywait(tm->ftm_fab->fab_fab, tm->ftm_fids.ftf_head,
+				 tm->ftm_fids.ftf_cnt);
 		/*
-		 * The lab team observed a significant increase in the CPU load
-		 * due to the epoll_wait not sleeping.
-		 * The proposed short term fix to support the lab team is adding
-		 * nanosleep of 0ns on VM to reduce CPU load.
-		 */
-		if (m0_processor_is_vm())
-			m0_nanosleep(M0_MKTIME(0 ,0), NULL);
-		/*
-		 * It is observed that with epoll_wait,
-		 * the thread is waiting in a busy-loop for events
-		 * thus not releasing CPU.
-		 * Hence, adding a sched_yield() will release the CPU for
-		 * other processes and reduce CPU consumption.
+		 * TBD : Add handling of other return values of fi_trywait() if
+		 * it returns something other than -EAGAIN and 0.
 		 */
-		sched_yield();
-		ev_cnt = epoll_wait(tm->ftm_epfd, &ev, 1, FAB_WAIT_FD_TMOUT);
+		M0_ASSERT(M0_IN(ret, (0, -EAGAIN)));
```

Comment:
1000            while (tm->ftm_state != FAB_TM_SHUTDOWN) {
1001                    ret = fi_trywait(tm->ftm_fab->fab_fab, tm->ftm_fids.ftf_head,
1002                                     tm->ftm_fids.ftf_cnt);
(gdb)
1003                    /*
1004                     * TBD : Add handling of other return values of fi_trywait() if
1005                     * it returns something other than -EAGAIN and 0.
1006                     */
1007                    M0_ASSERT(M0_IN(ret, (0, -EAGAIN)));
1008                    if (ret == 0) {
1009                            err = -EINTR;
1010                            while (err == -EINTR) {
1011                                    ret = epoll_wait(tm->ftm_epfd, &ev, 1,
1012                                                     FAB_WAIT_FD_TMOUT);
(gdb) bt
#0  0x00007f44458753d7 in raise () from /lib64/libc.so.6
#1  0x00007f4445876ac8 in abort () from /lib64/libc.so.6
#2  0x00007f4447945a80 in m0_arch_panic (c=0x7f4447f402e0 <__pctx.12732>, ap=0x7f44093f1c28)
    at lib/user_space/uassert.c:131
#3  0x00007f4447925ff6 in m0_panic (ctx=0x7f4447f402e0 <__pctx.12732>) at lib/assert.c:52
#4  0x00007f44479e07ba in libfab_poller (tm=0x1e06e80) at net/libfab/libfab.c:1007
#5  0x00007f4447933499 in m0_thread_trampoline (arg=0x1e06e88) at lib/thread.c:117
#6  0x00007f444794715e in uthread_trampoline (arg=0x1e06e88) at lib/user_space/uthread.c:98
#7  0x00007f4446fbeea5 in start_thread () from /lib64/libpthread.so.0
#8  0x00007f444593d9fd in clone () from /lib64/libc.so.6
(gdb) p ret
$2 = -22
(gdb)

With 41* ST on top of kubernetes branch i have got -22 error here.

Reply:
Observed that fi_trywait() returns -22(EINVAL) which is not mentioned in libfabric documentation, hence added it to the list of possible return values of fi_trywait().